src/shogun/machine/GLM.h

```diff
@@ -0,0 +1,268 @@
+		
+		/** default constructor */
+		GLM();
+
+ 		GLM(GLM_DISTRIBUTION distribution, float64_t alpha, float64_t lambda, float64_t learning_rate, int32_t max_iterations, float64_t tolerance, float64_t eta);
+
+ 		/** destructor */
+ 		~GLM() override {}
+
+		/** apply linear machine to data
+		 * for regression problem
+		 *
+		 * @param data (test)data to be classified
+		 * @return classified labels
+		 */
+		std::shared_ptr<RegressionLabels> apply_regression(std::shared_ptr<Features> data=NULL) override;
+
+ 		/** Returns the name of the SGSerializable instance.  It MUST BE
+ 		 *  the CLASS NAME without the prefixed `C'.
+ 		 *
+ 		 * @return name of the SGSerializable
+ 		 */
+ 		const char* get_name() const override { return "GLM"; }
+
+		void set_tau(SGMatrix<float64_t> tau);
+
+		SGMatrix<float64_t> get_tau();
+
+	protected:
+
+		void init_model(const std::shared_ptr<Features>& data) override;
+		
+		void iteration() override;
+
+	protected:
+
+		/** Distribution type */
+		GLM_DISTRIBUTION distribution = POISSON;
+
+		/** a threshold parameter that linearizes the exp() function above eta. */
+		float64_t m_eta = 2.0;
+
+		/** regularization parameter :math:`\\lambda` of penalty term. */
+		float64_t m_lambda = 0.1;
+
+		/** the weighting between L1 penalty and L2 penalty term of the loss function. */
+		float64_t m_alpha = 0.5;
+
+		/** convergence threshold or stopping criteria. Optimization loop will stop when relative change in parameter norm is below the threshold. */
+		float64_t m_tolerance = 1e-6;
+
+		bool m_compute_bias = true;
+
+		std::shared_ptr<GradientDescendUpdater> m_gradient_updater;
+
+		std::shared_ptr<ConstLearningRate> m_learning_rate;
+
+		std::shared_ptr<ElasticNetPenalty> m_penalty;
+
+		std::shared_ptr<GLMCostFunction> m_cost_function;
+};
+
+class GLMCostFunction: public FirstOrderCostFunction
+{
+public:
+
+	friend class GLM;
+
+	void set_target(const std::shared_ptr<GLM>&obj)
+	{
+		if(m_obj != obj)
+			m_obj=obj;
+	}
+
+	void unset_target()
+	{
+		m_obj=NULL;
+	}
+
+	virtual float64_t get_cost()
+	{
+		//TODO
+		return 0.0;
+	}
+
+	virtual SGVector<float64_t> obtain_variable_reference()
+	{
+		require(m_obj,"Object not set");
+		return m_obj->m_w;
+	}
+
+	virtual SGVector<float64_t> get_gradient()
+	{
+		auto X = m_obj->LinearMachine::features->get_computed_dot_feature_matrix();
+		auto y = m_obj->m_labels->get_values();
+		SGVector<float64_t> w_old(m_obj->m_w);
+
+		auto n_samples = y.vlen;
+	
+		auto z = compute_z(X, m_obj->m_w, m_obj->bias);
+		auto mu = non_linearity(z);
+		auto grad_mu = gradient_non_linearity(z);
+
+		SGVector<float64_t> grad_w(m_obj->m_w.vlen);
+		SGVector<float64_t> a;
+		switch (m_obj->distribution)
+		{
+		case POISSON:
+			a = linalg::element_prod(y, grad_mu);
+			for(int i = 0; i<y.vlen; i++)
+				a[i] /= mu[i];
+			linalg::transpose_matrix(linalg::add(linalg::matrix_prod(SGMatrix(grad_mu), X, true, false), (linalg::matrix_prod(SGMatrix(a), X, true, false)), 1.0, -1.0));
```

Comment:
I think in this case it wouldn't hurt to break up this line into several lines (you won't miss out on any performance) and the code will be more readable

Reply:
Yes, I'll rectify that.